hw/ip/otbn/dv/rig/rig/gens/misaligned_load_store.py

```diff
@@ -0,0 +1,290 @@
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        weights = self.weights
+        prog_insn = None
+        while prog_insn is None:
+            idx = random.choices(range(len(self.insns)), weights=weights)[0]
+
+            # If all 4 instructions are not successfull, give up.
+            if not weights[idx]:
+                return None
+
+            # Try to fill out the instruction. On failure, clear the weight for
+            # this index and go around again.
+            prog_insn = self.fill_insn(self.insns[idx], model)
+            if prog_insn is None:
+                weights[idx] = 0
+                continue
+
+        weights = self.weights.copy()
+        snippet = ProgSnippet(model.pc, [prog_insn])
+        snippet.insert_into_program(program)
+
+        return (snippet, True, model)
+
+    def fill_insn(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+        '''Try to pick one of BN.XID or XW instructions
+
+        '''
+
+        # Special-case BN load/store instructions by mnemonic. These use
+        # complicated indirect addressing, so it's probably more sensible to
+        # give them special code.
+        if insn.mnemonic in ['bn.lid', 'bn.sid']:
+            return self._fill_bn_xid(insn, model)
+        if insn.mnemonic in ['lw', 'sw']:
+            return self._fill_xw(insn, model)
+
+        return None
+
+    def _fill_xw(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+
+        grd_op_type = self.lw.operands[0].op_type
+        imm_op_type = self.lw.operands[1].op_type
+
+        # Determine the offset range for XW
+        offset_rng = imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        # Max/Min Offsets for XW (-2048, 2047)
+        min_offset, max_offset = offset_rng
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+
+        base = []
+        for reg_idx, reg_val in known_regs:
+            val = reg_val - (1 << 32) if reg_val << 31 else reg_val
+            if (1 - max_offset <= val < model.dmem_size - min_offset):
+                base.append((reg_idx, val))
+
+        assert base is not None
+
+        # Pick a random register among known registers that is constrained above
+        idx, value = random.choice(base)
+
+        op_val_grs1 = idx
+        assert op_val_grs1 is not None
+
+        # Pick a random offset value. We will change it later for guaranteeing
+        # misaligned addresses.
+        imm_val = random.randrange(min_offset, max_offset)
+
+        # Initialize the address value for the while loop below
+        addr = imm_val + value
+
+        # Randomize immediate until getting a misaligned address which is in a
+        # valid range otherwise
+        while ((not addr % 4) or addr < 0 or model.dmem_size <= addr):
+            imm_val = random.choice(range(min_offset, max_offset))
+            addr = imm_val + value
```

Comment:
You can avoid the repetition by re-casting things slightly. I think this is probably clearer?
```suggestion
        imm_val = None
        for _ in range(50):
            # Pick a random offset value. We will change it later for
            # guaranteeing misaligned addresses.
            imm_val = random.randrange(min_offset, max_offset)
            addr = imm_val + value
            if (addr % 4) and (0 <= addr < model.dmem_size):
                break

        # The loop body above should succeed 75% of the time, so we should only
        # get here without an imm_val one time in 2^200.
        assert imm_val is not None
```